### PR TITLE
UNR-757 Detect out of order rpc

### PR DESCRIPTION
**Pull request by improbable-valentyn**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
We need to guarantee that reliable RPCs are executed in the order they are called (on the same actor), but first we need to know when RPCs are executed out of order.
This PR accomplishes that by attaching an index with the RPC so the receiving side can detect when the index of the RPC about to be executed is different from the one it expects.
#### Tests
Ran ShooterGame with this, which already revealed out-of-order reliable RPCs caused by one argument being an unresolved object (Client RPCs on PlayerController waiting for the Pawn).
#### Documentation
https://improbableio.atlassian.net/browse/UNR-757
#### Primary reviewers
@m-samiec @joshuahuburn

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialWorkerConnection.cpp [hunk 1]
```diff
@@ -325,5 +325,5 @@ void USpatialWorkerConnection::CacheWorkerAttributes()
 	for (uint32 Index = 0; Index < Attributes->attribute_count; ++Index)
 	{
 		CachedWorkerAttributes.Add(UTF8_TO_TCHAR(Attributes->attributes[Index]));
-	}	
+	}
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialWorkerConnection.cpp]: @danielimprobable

Comment (danielimprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialWorkerConnection.cpp]: WHYYYYYYYYYYYYYYYYYYY, *in tears** #Sorry NOOOOOOOO

File: SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialReceiver.h [hunk 2]
```diff
@@ -86,6 +86,9 @@ struct FPendingIncomingRPC
 	UFunction* Function;
 	TArray<uint8> PayloadData;
 	int64 CountBits;
+#if !UE_BUILD_SHIPPING
+	FString SenderWorkerId;
+#endif
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialReceiver.h]:
```suggestion
#endif // !UE_BUILD_SHIPPING
```

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h [hunk 3]
```diff
@@ -127,6 +127,25 @@ class SPATIALGDK_API USpatialNetDriver : public UIpNetDriver
...
+		FReliableRPCId(FString InWorkerId, uint32 InRPCId, FString InRPCTarget, FString InRPCName) : WorkerId(InWorkerId), RPCId(InRPCId), LastRPCTarget(InRPCTarget), LastRPCName(InRPCName) {}
+
+		FString WorkerId;
+		uint32 RPCId = 0;
+		FString LastRPCTarget;
+		FString LastRPCName;
+	};
+	using FRPCTypeToReliableRPCIdMap = TMap<ESchemaComponentType, FReliableRPCId>;
+	// Per actor, maps from RPC type to the reliable RPC index used to detect if reliable RPCs go out of order.
+	TMap<TWeakObjectPtr<AActor>, FRPCTypeToReliableRPCIdMap> ReliableRPCIdMap;
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h]: Currently, this will leak. We could swipe it for deleted actors on tick, or once in a while, but I figured that may be a premature optimization for a debug functionality.

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h [hunk 4]
```diff
@@ -67,12 +67,12 @@ class SPATIALGDK_API USpatialActorChannel : public UActorChannel
 		const FClassInfo* Info = NetDriver->TypebindingManager->FindClassInfoByClass(Actor->GetClass());
 		check(Info);
 
-		const TArray<FString> WorkerAttributes = NetDriver->Connection->GetWorkerAttributes();
+		const TArray<FString>& WorkerAttributes = NetDriver->Connection->GetWorkerAttributes();
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h]: Drive-by changes to the code from https://github.com/spatialos/UnrealGDK/pull/570. I could move this to a separate PR if anyone feels strongly about it.

File: SpatialGDK/Source/SpatialGDK/Public/Interop/Connection/SpatialWorkerConnection.h [hunk 5]
```diff
@@ -39,7 +39,7 @@ class SPATIALGDK_API USpatialWorkerConnection : public UObject
 	void SendLogMessage(const uint8_t Level, const char* LoggerName, const char* Message);
 	void SendComponentInterest(Worker_EntityId EntityId, const TArray<Worker_InterestOverride>& ComponentInterest);
 	FString GetWorkerId() const;
-	TArray<FString> GetWorkerAttributes() const;
+	const TArray<FString>& GetWorkerAttributes() const;
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/Connection/SpatialWorkerConnection.h]: Drive-by changes to the code from https://github.com/spatialos/UnrealGDK/pull/570. I could move this to a separate PR if anyone feels strongly about it.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 6]
```diff
@@ -1375,3 +1384,91 @@ void USpatialNetDriver::WipeWorld(const USpatialNetDriver::PostWorldWipeDelegate
 		SnapshotManager->WorldWipe(LoadSnapshotAfterWorldWipe);
 	}
 }
+
+#if !UE_BUILD_SHIPPING
+uint8 USpatialNetDriver::GetNextReliableRPCId(AActor* Actor, ESchemaComponentType RPCType, UObject* TargetObject)
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]:
```suggestion
uint32 USpatialNetDriver::GetNextReliableRPCId(AActor* Actor, ESchemaComponentType RPCType, UObject* TargetObject)
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 7]
```diff
@@ -1375,3 +1384,91 @@ void USpatialNetDriver::WipeWorld(const USpatialNetDriver::PostWorldWipeDelegate
...
+	else
+	{
+		// Add an entry for this RPC type with empty WorkerId and RPCId = 0
+		ReliableRPCIds.Add(RPCType);
+	}
+
+	return ++ReliableRPCIds[RPCType].RPCId;
+}
+
+void USpatialNetDriver::OnReceivedReliableRPC(AActor* Actor, ESchemaComponentType RPCType, FString WorkerId, uint8 RPCId, UObject* TargetObject, UFunction* Function)
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]:
```suggestion
void USpatialNetDriver::OnReceivedReliableRPC(AActor* Actor, ESchemaComponentType RPCType, FString WorkerId, uint32 RPCId, UObject* TargetObject, UFunction* Function)
```

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h [hunk 8]
```diff
@@ -127,6 +127,25 @@ class SPATIALGDK_API USpatialNetDriver : public UIpNetDriver
 	void StartIgnoringAuthoritativeDestruction() { bAuthoritativeDestruction = false; }
 	void StopIgnoringAuthoritativeDestruction() { bAuthoritativeDestruction = true; }
 
+#if !UE_BUILD_SHIPPING
+	uint8 GetNextReliableRPCId(AActor* Actor, ESchemaComponentType RPCType, UObject* TargetObject);
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h]:
```suggestion
	uint32 GetNextReliableRPCId(AActor* Actor, ESchemaComponentType RPCType, UObject* TargetObject);
```

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h [hunk 9]
```diff
@@ -127,6 +127,25 @@ class SPATIALGDK_API USpatialNetDriver : public UIpNetDriver
 	void StartIgnoringAuthoritativeDestruction() { bAuthoritativeDestruction = false; }
 	void StopIgnoringAuthoritativeDestruction() { bAuthoritativeDestruction = true; }
 
+#if !UE_BUILD_SHIPPING
+	uint32 GetNextReliableRPCId(AActor* Actor, ESchemaComponentType RPCType, UObject* TargetObject);
+	void OnReceivedReliableRPC(AActor* Actor, ESchemaComponentType RPCType, FString WorkerId, uint8 RPCId, UObject* TargetObject, UFunction* Function);
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h]:
```suggestion
	void OnReceivedReliableRPC(AActor* Actor, ESchemaComponentType RPCType, FString WorkerId, uint32 RPCId, UObject* TargetObject, UFunction* Function);
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 10]
```diff
@@ -822,12 +822,12 @@ void USpatialReceiver::ReceiveMulticastUpdate(const Worker_ComponentUpdate& Comp
...
 			int64 CountBits = PayloadData.Num() * 8;
 
-			ApplyRPC(TargetObject, Function, PayloadData, CountBits);
+			ApplyRPC(TargetObject, Function, PayloadData, CountBits, FString());
 		}
 	}
 }
 
-void USpatialReceiver::ApplyRPC(UObject* TargetObject, UFunction* Function, TArray<uint8>& PayloadData, int64 CountBits)
+void USpatialReceiver::ApplyRPC(UObject* TargetObject, UFunction* Function, TArray<uint8>& PayloadData, int64 CountBits, FString SenderWorkerId)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]:
```suggestion
void USpatialReceiver::ApplyRPC(UObject* TargetObject, UFunction* Function, TArray<uint8>& PayloadData, int64 CountBits, const FString& SenderWorkerId)
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 11]
```diff
@@ -991,9 +1011,12 @@ void USpatialReceiver::QueueIncomingRepUpdates(FChannelObjectPair ChannelObjectP
 	}
 }
 
-void USpatialReceiver::QueueIncomingRPC(const TSet<FUnrealObjectRef>& UnresolvedRefs, UObject* TargetObject, UFunction* Function, const TArray<uint8>& PayloadData, int64 CountBits)
+void USpatialReceiver::QueueIncomingRPC(const TSet<FUnrealObjectRef>& UnresolvedRefs, UObject* TargetObject, UFunction* Function, const TArray<uint8>& PayloadData, int64 CountBits, FString SenderWorkerId)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]:
```suggestion
void USpatialReceiver::QueueIncomingRPC(const TSet<FUnrealObjectRef>& UnresolvedRefs, UObject* TargetObject, UFunction* Function, const TArray<uint8>& PayloadData, int64 CountBits, const FString& SenderWorkerId)
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 12]
```diff
@@ -1210,13 +1237,13 @@ void USpatialReceiver::ResolveObjectReferences(FRepLayout& RepLayout, UObject* R
 	}
 }
 
-void USpatialReceiver::ReceiveRPCCommandRequest(const Worker_CommandRequest& CommandRequest, UObject* TargetObject, UFunction* Function)
+void USpatialReceiver::ReceiveRPCCommandRequest(const Worker_CommandRequest& CommandRequest, UObject* TargetObject, UFunction* Function, FString SenderWorkerId)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]:
```suggestion
void USpatialReceiver::ReceiveRPCCommandRequest(const Worker_CommandRequest& CommandRequest, UObject* TargetObject, UFunction* Function, const FString& SenderWorkerId)
```

File: SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialReceiver.h [hunk 13]
```diff
@@ -144,14 +147,14 @@ class USpatialReceiver : public UObject
 	void ApplyComponentData(Worker_EntityId EntityId, Worker_ComponentData& Data, USpatialActorChannel* Channel);
 	void ApplyComponentUpdate(const Worker_ComponentUpdate& ComponentUpdate, UObject* TargetObject, USpatialActorChannel* Channel, bool bIsHandover);
 
-	void ReceiveRPCCommandRequest(const Worker_CommandRequest& CommandRequest, UObject* TargetObject, UFunction* Function);
+	void ReceiveRPCCommandRequest(const Worker_CommandRequest& CommandRequest, UObject* TargetObject, UFunction* Function, FString SenderWorkerId);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialReceiver.h]:
```suggestion
	void ReceiveRPCCommandRequest(const Worker_CommandRequest& CommandRequest, UObject* TargetObject, UFunction* Function, const FString& SenderWorkerId);
```

File: SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialReceiver.h [hunk 14]
```diff
@@ -144,14 +147,14 @@ class USpatialReceiver : public UObject
...
 	void ReceiveMulticastUpdate(const Worker_ComponentUpdate& ComponentUpdate, UObject* TargetObject, const TArray<UFunction*>& RPCArray);
-	void ApplyRPC(UObject* TargetObject, UFunction* Function, TArray<uint8>& PayloadData, int64 CountBits);
+	void ApplyRPC(UObject* TargetObject, UFunction* Function, TArray<uint8>& PayloadData, int64 CountBits, FString SenderWorkerId);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialReceiver.h]:
```suggestion
	void ApplyRPC(UObject* TargetObject, UFunction* Function, TArray<uint8>& PayloadData, int64 CountBits, const FString& SenderWorkerId);
```

File: SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialReceiver.h [hunk 15]
```diff
@@ -144,14 +147,14 @@ class USpatialReceiver : public UObject
...
 
 	void ReceiveCommandResponse(Worker_CommandResponseOp& Op);
 
 	void QueueIncomingRepUpdates(FChannelObjectPair ChannelObjectPair, const FObjectReferencesMap& ObjectReferencesMap, const TSet<FUnrealObjectRef>& UnresolvedRefs);
-	void QueueIncomingRPC(const TSet<FUnrealObjectRef>& UnresolvedRefs, UObject* TargetObject, UFunction* Function, const TArray<uint8>& PayloadData, int64 CountBits);
+	void QueueIncomingRPC(const TSet<FUnrealObjectRef>& UnresolvedRefs, UObject* TargetObject, UFunction* Function, const TArray<uint8>& PayloadData, int64 CountBits, FString SenderWorkerId);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialReceiver.h]:
```suggestion
	void QueueIncomingRPC(const TSet<FUnrealObjectRef>& UnresolvedRefs, UObject* TargetObject, UFunction* Function, const TArray<uint8>& PayloadData, int64 CountBits, const FString& SenderWorkerId);
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 16]
```diff
@@ -1375,3 +1384,91 @@ void USpatialNetDriver::WipeWorld(const USpatialNetDriver::PostWorldWipeDelegate
...
+	else
+	{
+		// Add an entry for this RPC type with empty WorkerId and RPCId = 0
+		ReliableRPCIds.Add(RPCType);
+	}
+
+	return ++ReliableRPCIds[RPCType].RPCId;
+}
+
+void USpatialNetDriver::OnReceivedReliableRPC(AActor* Actor, ESchemaComponentType RPCType, FString WorkerId, uint32 RPCId, UObject* TargetObject, UFunction* Function)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Early out if WorkerId is empty?
